subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java

```diff
@@ -16,161 +16,228 @@
     private final ProjectPropertiesLoader projectPropertiesLoader;
 
-    public DefaultGradlePropertiesController(IGradlePropertiesLoader propertiesLoader, SystemPropertiesInstaller systemPropertiesInstaller, ProjectPropertiesLoader projectPropertiesLoader) {
+    private final ConcurrentMap<BuildIdentifier, BuildScopedGradleProperties> buildProperties = new ConcurrentHashMap<>();
+    private final ConcurrentMap<ProjectIdentity, ProjectScopedGradleProperties> projectProperties = new ConcurrentHashMap<>();
+
+    public DefaultGradlePropertiesController(
+        Environment environment,
+        IGradlePropertiesLoader propertiesLoader,
+        SystemPropertiesInstaller systemPropertiesInstaller,
+        ProjectPropertiesLoader projectPropertiesLoader
+    ) {
+        this.environment = environment;
         this.propertiesLoader = propertiesLoader;
         this.systemPropertiesInstaller = systemPropertiesInstaller;
         this.projectPropertiesLoader = projectPropertiesLoader;
     }
 
     @Override
-    public GradleProperties getGradleProperties() {
-        return sharedGradleProperties;
+    public GradleProperties getGradleProperties(BuildIdentifier buildId) {
+        return getOrCreateGradleProperties(buildId);
     }
 
     @Override
-    public void loadGradlePropertiesFrom(File settingsDir, boolean setSystemProperties) {
-        state = state.loadGradlePropertiesFrom(settingsDir, setSystemProperties);
+    public void loadGradleProperties(BuildIdentifier buildId, File buildRootDir, boolean setSystemProperties) {
+        getOrCreateGradleProperties(buildId).loadProperties(buildRootDir, setSystemProperties);
     }
 
     @Override
-    public void unloadGradleProperties() {
-        state = new NotLoaded();
+    public void unloadGradleProperties(BuildIdentifier buildId) {
+        getOrCreateGradleProperties(buildId).unload();
     }
 
-    public void overrideWith(GradleProperties gradleProperties) {
-        state = state.overrideWith(gradleProperties);
+    @Override
+    public GradleProperties getGradleProperties(ProjectIdentity projectId) {
+        return getOrCreateGradleProperties(projectId);
     }
 
-    private class SharedGradleProperties implements GradleProperties {
+    @Override
+    public void loadGradleProperties(ProjectIdentity projectId, File projectDir) {
+        getOrCreateGradleProperties(projectId).loadProperties(projectDir, false);
+    }
+
+    private BuildScopedGradleProperties getOrCreateGradleProperties(BuildIdentifier buildId) {
+        return buildProperties.computeIfAbsent(buildId, BuildScopedGradleProperties::new);
+    }
+
+    private ProjectScopedGradleProperties getOrCreateGradleProperties(ProjectIdentity projectId) {
+        return projectProperties.computeIfAbsent(projectId, ProjectScopedGradleProperties::new);
+    }
+
+    private abstract static class SharedGradleProperties<I, T extends GradleProperties> implements GradleProperties {
```

Comment:
🤔 technically, we don't unload project properties, so YAGNI tells us that the unload support in base class isn't needed. But It doesn't add much mental overhead anyway.

Reply:
We'll probably polish this further in follow-up PRs